### PR TITLE
Fix Expander in CollectionView on Apple

### LINKED ISSUE #742: [BUG] Add `CollectionView` and `ListView` Support to `Expander`
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When `Expander` is used inside of a `CollectionView` or a `ListView`, a `NotSupportedException` is thrown:
https://github.com/CommunityToolkit/Maui/blob/1ea488c9be7aac234af32e1ab5b014d6a52a5ff0/src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs#L197-L202

### Expected Behavior

When an `Expander` is expanded, its parent controls should re-measure their size and layouts.

Here is a incomplete sample at how we could approach solving this problem:

```cs
if (element is ListView listView)
{
  foreach (var child in listView.AllChildren.OfType<Cell>())
  {
	  child.ForceUpdateSize();
  }
  listView.InvalidateMeasureInternal(Microsoft.Maui.Controls.Internals.InvalidationTrigger.MeasureChanged);
}
else if (element is CollectionView collectionView)
{
  collectionView.InvalidateMeasureInternal(Microsoft.Maui.Controls.Internals.InvalidationTrigger.MeasureChanged);
}
```

### Steps To Reproduce

1. Download/Clone the [Expander-ItemsView-NotSupportedException branch in `CommunityToolkit.Maui`](https://github.com/CommunityToolkit/Maui/tree/Expander-ItemsView-NotSupportedException)
2. In Visual Studio, open `CommunityToolkit.Maui.Samples.sln`
3. Navigate to the [`ExpanderPage`](https://github.com/CommunityToolkit/Maui/blob/d93f2aa8cda296de2ae39e9c294b891065c7a6bc/samples/CommunityToolkit.Maui.Sample/Pages/Views/Expander/ExpanderPage.xaml#L51-L96)
4. Confirm the app crashes when the ExpanderPage opens

### Link to public reproduction project repository

https://github.com/CommunityToolkit/Maui/blob/d93f2aa8cda296de2ae39e9c294b891065c7a6bc/samples/CommunityToolkit.Maui.Sample/Pages/Views/Expander/ExpanderPage.xaml#L51-L96

### Environment

```markdown
- .NET MAUI CommunityToolkit: v1.4.0
- OS: All
- .NET MAUI: 6.0.312
```

### Anything else?

This is a known-issue, documented in the .NET MAUI Community Toolkit documentation: https://learn.microsoft.com/dotnet/communitytoolkit/maui/views/expander

### PR DESCRIPTION
**Pull request by VladislavAntonyuk**
### Description of Change ###

Set valid size and invalidate layout of CollectionView

 ### Linked Issues ###
 

 - Fixes #742 

 ### PR Checklist ###
 
 - [ ] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [ ] Has samples (if omitted, state reason in description)
 - [ ] Rebased on top of `main` at time of PR
 - [ ] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/XYZ 

 ### Additional information ###

### CODE REVIEW

File: src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs
Comment (inforithmics) [inline src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:248]: Isn't there a code path needed for Android Windows
#elif ANDROID ....
#elif WINDOWS ....

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:248]: other platforms seems to work correct without this additional recalculation

File: src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs
Comment (pictos) [inline src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:236]: IMHO this doesn't look very solid. I would say the best solution is to work with someone .NET Maui team and try to expose the property that we need or see if there's other way to grab this information

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:236]: It is protected: `protected ItemsViewController<TItemsView> Controller { get; }`
image

Comment (pictos) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:236]: So let's talk with the team about it. Any chance that you can send the link to the Maui code line that we need to have access to?

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:236]: src/Controls/src/Core/Handlers/Items/ItemsViewHandler.iOS.cs

https://github.com/dotnet/maui/blob/05ad340fd6b5e8d673da21642c4bdc178f890554/src/Controls/src/Core/Handlers/Items/ItemsViewHandler.iOS.cs#L37-L38

Comment (brminnick) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:238]: @pictos / @VladislavAntonyuk Any progress on getting the needed code promoted to `public` or `internal` via the .NET MAUI engineering team?

Comment (pictos) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:238]: no progress so far

**Comment by pictos** [comment]
I tried on my end, and that didn't fix for maccatalyst, as you can see in the video below

https://user-images.githubusercontent.com/20712372/209261513-f835e72d-fb02-4141-b5f1-e5d2facbcdc4.mp4

**Comment by VladislavAntonyuk** [comment]
Thank you @pictos. Seems like Task.Delay before InvalidateLayout was not a good solution, that works on all devices. I reworked it. Please check if it works for you

**Comment by rmarinho** [comment]
We merged the change on MAUI

**Comment by VladislavAntonyuk** [comment]
> We merged the change on MAUI

Thank you @rmarinho! Will it be included in the next release?

**Comment by rmarinho** [comment]
I don't think so, unless we backport the fix to net7

**Comment by pictos** [comment]
@VladislavAntonyuk before moving with this and the other PR that I opened, the fix didn't work well for ListView on maccatalyst

**Comment by VladislavAntonyuk** [comment]
@pictos the fix is only for CollectionView. I also saw back merge of .NET MAUI to avoid reflection. 
I believe we should wait for that release to avoid the workarounds

File: src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs
Comment (brminnick) [inline src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: @pictos @VladislavAntonyuk 

FYI - I simplified the code from #890 a bit, removing the `public static Action`, replacing it with `public virtual void OnHeaderTapped()`.

I configured Expander to [run `OnHeaderTapped()` every time `Expander.tapGestureRecognizer` is triggered](https://github.com/CommunityToolkit/Maui/pull/751/files#diff-a7933be10a281abb7a9062799e5eb3b5e633518aaff6326c9a8d1b4adc829a6bR251).

This ensures that `Expander` supports `CollectionView` by default, and it allows users to override this method, implementing their own logic if they'd like. It also allows us to one day remove the code inside `OnHeaderTapped()` if/when .NET MAUI fixes the bug on their end without causing any breaking changes to MCT devs.

Comment (brminnick) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: We should also add a Unit Test for this new method since it is `public`, if we can

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: With public virtual method user needs to inherit from Expander and create his own control.
With public static property he only need to create an action and assign it.
Public static property seems to be a simpler solution from user perspective. We can mark that property obsolete and remove it with major release.

Comment (brminnick) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: > With public virtual method user needs to inherit from Expander and create his own control. With public static property he only need to create an action and assign it. Public static property seems to be a simpler solution from user perspective

True, but let's zoom out a bit: How many developers are actually going to change the default behavior? My guess is not many, if any. 

If a developer is going to go through the trouble of changing the default behavior of `Expander`, I doubt they'll mind creating a custom Expander class. It doesn't require many lines of code:

```cs
class CustomExpander : Expander
{
  public virtual void OnHeaderTapped(TappedEventArgs tappedEventArgs)
  {
    base.OnHeaderTapped(tappedEventArgs);
    
    // custom code
  }
}
```

> We can mark that property obsolete and remove it with major release

My goal is to avoid us purposely adding a property that we knowingly will `[Obsolete]` later because `[Obsolete]` code adds technical debt and becomes a headache to manage in the code + the docs.

tl;dr - I prefer that we be selfish and avoid purposely adding technical debt, even if it requires a small bit more work for the developer to create a new subclass to override the default behavior

Comment (pictos) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: @brminnick, I would recommend keeping the `Action` approach, take a sample of how you can customize the PlatformView on .NET MAUI `EntryHandler.PlatformFactoryView = (handler) => return PlatformView;`, the idea .NET MAUI is to avoid inheritance and have external points of extensibility.

Comment (pictos) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: > My goal is to avoid us purposely adding a property that we knowingly will [Obsolete] later because [Obsolete] code adds technical debt and becomes a headache to manage in the codebase + the docs.

We don't need to Obsolete that, we can make it available for anyone that want to influence the Expand behavior, we will need to Obsolete the default methods when that be fixed.

Comment (brminnick) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: Let's again zoom out: Why are we even adding a public API? Are devs demanding this? 

If not, let's make `OnHeaderTapped` a private method.

I don't think many, if any, developers are going to modify the default functionality.

Comment (pictos) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: Sorry, but I still can't see your point be an argument against the usage of an Action instead of `virtual` method.

Comment (brminnick) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: My question is Why are we adding this new API? Are developers asking for this? Do we have an open Issue for this?

I understand it's a workaround to adding Expander support to CollectionView + ListView. But, this PR is created to _fix_ Expander support for CollectionView + ListView, right?

So once this PR is merged, we won't need this new API, right?

Comment (pictos) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: This PR, IMHO, doesn't **fix** the issue, it provides a workaround since the fix should be in the .NET MAUI. We merged a PR on MAUI to improve the workaround (for .net 8), making it possible to get access to the members that we need without reflection.
So, since this is a workaround devs can opt-in to use ours or create their own implementation, with that we don't have the _obligation_ to support all possible scenarios (I would imagine there's at least one scenario that our workaround will fail) so let's say we provide a fix for 80% of the cases and the other 20% will have an easy way to implement their own fix.

This new API (Action) is a point of extensibility that devs can use to implement their own delegate in other to control that behavior, the inspiration come from [here](https://github.com/dotnet/maui/blob/3bae70081e3bb2639cf558e8351454368ac811c5/src/Core/src/Handlers/View/ViewHandlerOfT.cs#L58), you can see the use at line 72 of the same file. The idea is to have a place that devs can customize without need to extend the control, what's normal on XF but not on .NET MAUI

Comment (brminnick) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:137]: Ah - got it. 

Thanks man - I remember the conversation from our standup now. 

I'll clean up my mess in the morning 👍

Comment (brminnick) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs]: ✅ `public virtual void` reverted to `void`
✅ `Action` added with updated XML documentation

File: src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs
Comment (pictos) [inline src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:85]: this should be static or the dev will need to implement it for each instance of `Expander`
see 
- https://github.com/CommunityToolkit/Maui/blob/58240c8109f8d1bfe8313ddb4dc81898671ee728/src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs#L255

Comment (brminnick) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:85]: True, but I think it's best to keep it non-static. 

If a dev is using multiple `Expander`s in their app, they may not want to override the behavior the same way for each. 

And if a dev_does want to implement the same behavior for every `Expander`, they can subclass it like so:

```cs
public class CustomExpander : Expander
{
  public CustomExpander()
  {
    this.HandleHeaderTapped = CustomHeaderTappedMethod;
  }
  
  void CustomHeaderTappedMethod(TappedEventArgs args)
  {
    // ...
  }
}
```

Comment (pictos) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:85]: @brminnick  Can you double-check if the sample is still working? Since we use a collection, I believe we will create one per cell.

Comment (brminnick) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:85]: Of course! 

Yup, here is `Expander` inside of a `CollectionView`, working on iOS + MacCatalyst:
image

File: src/CommunityToolkit.Maui/Views/Expander/Expander.macios.cs
Comment (pictos) [inline src/CommunityToolkit.Maui/Views/Expander/Expander.macios.cs:31]: is there a change to `i` be less than zero?

File: src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs
Comment (pictos) [inline src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:239]: @VladislavAntonyuk you mentioned [here](https://github.com/CommunityToolkit/Maui/pull/751#issuecomment-1380883728) this PR just fix for `CollectionView`, so why do we have a `ListView` handle here?

**Review by pictos — CHANGES_REQUESTED** [review]
I've two suggestions. I just focused on the `Expander` related changed

File: src/CommunityToolkit.Maui/Views/Expander/Expander.macios.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.macios.cs:31]: i can be equal to 0

File: src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Views/Expander/Expander.shared.cs:239]: because for listview we still need to throw an unsupported exception on Apple.
But on other platforms, we can use it with ForceUpdateSize().

**Review by pictos — APPROVED** [review]

**Comment by brminnick** [comment]
@VladislavAntonyuk Can you update the docs, changing it to "Expander is not supported inside `ListView`" now that only ListView is not supported?

image

**Comment by AnthonyNjuguna** [comment]
This doesn't work if the collection has a header view. 

`var clickedCell = GetCellByPoint(cells, new CGPoint(tapLocation.Value.X, tapLocation.Value.Y));` 

the clickedCell always returns null.

**Comment by pictos** [comment]
@AnthonyNjuguna can you please open a new bug issue?

**Pull request merged**
